daemon/containerd/image_delete.go

```diff
@@ -78,9 +140,206 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
+	}
+
+	return fmt.Sprintf("conflict: unable to delete %s (%s) - %s", stringid.TruncateID(idc.imgID.String()), forceMsg, idc.message)
+}
+
+func (idc *imageDeleteConflict) Conflict() {}
+
+// imageDeleteHelper attempts to delete the given image from this daemon. If
+// the image has any hard delete conflicts (child images or running containers
+// using the image) then it cannot be deleted. If the image has any soft delete
+// conflicts (any tags/digests referencing the image or any stopped container
+// using the image) then it can only be deleted if force is true. If the delete
+// succeeds and prune is true, the parent images are also deleted if they do
+// not have any soft or hard delete conflicts themselves. Any deleted images
+// and untagged references are appended to the given records. If any error or
+// conflict is encountered, it will be returned immediately without deleting
+// the image. If quiet is true, any encountered conflicts will be ignored and
+// the function will return nil immediately without deleting the image.
+func (i *ImageService) imageDeleteHelper(ctx context.Context, imgID image.ID, records *[]types.ImageDeleteResponseItem, force, prune, quiet bool) error {
+	// First, determine if this image has any conflicts. Ignore soft conflicts
+	// if force is true.
+	c := conflictHard
+	if !force {
+		c |= conflictSoft
+	}
+
+	if conflict := i.checkImageDeleteConflict(ctx, imgID, c); conflict != nil {
+		image, err := i.resolveImage(ctx, imgID.String())
+		if err != nil {
+			return err
+		}
+		if quiet && (!isDanglingImage(image) || conflict.used) {
+			// Ignore conflicts UNLESS the image is "dangling" or not being used in
+			// which case we want the user to know.
+			return nil
+		}
+
+		// There was a conflict and it's either a hard conflict OR we are not
+		// forcing deletion on soft conflicts.
+		return conflict
+	}
+
+	parent, err := i.Parent(ctx, imgID)
+	if err != nil {
+		parent = ""
+	}
+
+	refs, err := i.client.ImageService().List(ctx, "target.digest=="+imgID.String())
+	if err != nil {
+		return err
+	}
+	for _, eachRef := range refs {
+		err = i.client.ImageService().Delete(ctx, eachRef.Name, images.SynchronousDelete())
+		if err != nil {
+			return err
+		}
+		untaggedRef, err := reference.ParseAnyReference(eachRef.Name)
+		if err != nil {
+		}
+		*records = append(*records, types.ImageDeleteResponseItem{Untagged: reference.FamiliarString(untaggedRef)})
+	}
+
+	i.LogImageEvent(imgID.String(), imgID.String(), "delete")
+	*records = append(*records, types.ImageDeleteResponseItem{Deleted: imgID.String()})
+	if !prune || parent == "" {
+		return nil
+	}
+
+	// We need to prune the parent image. This means delete it if there are
+	// no tags/digests referencing it and there are no containers using it (
+	// either running or stopped).
+	// Do not force prunings, but do so quietly (stopping on any encountered
+	// conflicts).
+	return i.imageDeleteHelper(ctx, parent, records, false, true, true)
```

Comment:
I'm not sure if I understand why we're deleting the parent too. The children-parent relationship isn't real in the containerd store as they're not really linked in any way like they are in the graphdriver implementation.
No image depends on any other, we can safely delete any image and the containerd GC will delete the underlying content only if there are no other references to it.

Also, IIRC in this case:
```
$ docker run --name asdf -it ubuntu 
# do sth with the image ...
$ docker commit asdf my-image
$ docker rmi my-image
```

both ubuntu and my-image images would get deleted?

Reply:
Yeah, I noticed this while implementing it and ended up leaving it there since it could be useful in some cases, and also for discussion. In your example it wouldn't get removed, since `ubuntu` is tagged and we're only deleting the parent if the parent is dangling.
Still, it seems a bit "forced" and maybe it doesn't make sense to do this with the containerd store.

Reply:
@corhere thoughts?

Reply:
If `prune` is supposed to delete dangling parent images and dangling images are our way to ensure otherwise-unreferenced content is retained, don't we have to delete the dangling parent images to allow the content store to GC that content? Please correct any misconceptions I have; I have to admit I don't have a very firm grasp on the nuances of the Moby data model, nor of containerd's.

Reply:
Yeah, you've understood correctly -- this is one of those places where the difference in the data models is extra obvious and we have to try and find a reasonable middle ground. :sob:

In containerd's data model, we have images (more concretely, references to manifests) which might share layers, and that's the extent of the relationship between two "images".  In Moby's, images have parental chains, so we've got a lot of code here emulating those parental chains via layer sharing checks, which is very clever.

The situation this is handling is one that the Moby data model wouldn't get into - we build an image based on `ubuntu`, then we untag `ubuntu` (but not delete), and then we delete our new built image.  In Moby, that would delete the `ubuntu` content too.  In containerd, we have to track "dangling" content explicitly, so that untagged "ubuntu" ends up with a dangling reference that wouldn't have existed in the Moby data model because those layers would be a "child" of our newly built image and the "ubuntu" image wouldn't even show in `docker images` at all.

In typing this out, I think I've convinced myself that what Laura has implemented here _is_ correct and we should be pruning any dangling "parent" images when we delete the child, because that matches the Moby data model. :+1:

Reply:
<details>
<summary>A better illustration:</summary>

```console
$ docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
$ docker pull ubuntu
Using default tag: latest
latest: Pulling from library/ubuntu
2ab09b027e7f: Pull complete 
Digest: sha256:67211c14fa74f070d27cc59d69a7fa9aeff8e28ea118ef3babc295a0428a6d21
Status: Downloaded newer image for ubuntu:latest
docker.io/library/ubuntu:latest
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
ubuntu       latest    08d22c0ceb15   7 weeks ago   77.8MB
$ printf 'FROM ubuntu\nRUN touch foo' | DOCKER_BUILDKIT=0 docker build --tag foo -
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            BuildKit is currently disabled; enable it by removing the DOCKER_BUILDKIT=0
            environment-variable.

Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM ubuntu
 ---> 08d22c0ceb15
Step 2/2 : RUN touch foo
 ---> Running in 6570e399985c
Removing intermediate container 6570e399985c
 ---> f46a16d0b238
Successfully built f46a16d0b238
Successfully tagged foo:latest
$ docker rmi --no-prune ubuntu
Untagged: ubuntu:latest
Untagged: ubuntu@sha256:67211c14fa74f070d27cc59d69a7fa9aeff8e28ea118ef3babc295a0428a6d21
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
foo          latest    f46a16d0b238   16 seconds ago   77.8MB
$ docker rmi foo
Untagged: foo:latest
Deleted: sha256:f46a16d0b2380e977953a8f965670d86c49d5500cd455755d874cf527d4ffbbd
Deleted: sha256:a62c126ca0251fb4c4ac0b3685e54eafb0493d01f51e6f584af7615a6f137bc6
Deleted: sha256:08d22c0ceb150ddeb2237c5fa3129c0183f3cc6f5eeb2e7aa4016da3ad02140a
Deleted: sha256:b93c1bd012ab8fda60f5b4f5906bf244586e0e3292d84571d3abb56472248466
$ docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
```

</details>